### DISCUSSION TITLE
struct in a different assembly used without initialization

### DISCUSSION THREAD

**Discussion by sgurukrupa** [dotnet/csharplang#5230] [Q&A]
The following C# program doesn't compile

```
class Program
{
    public struct MetaWorkspace
    {
        public string Name { get; set; }
        public string Id { get; set; }
    }

    static void Main(string[] args)
    {
        MetaWorkspace ws;
        ws.Id = "abc";
    }
}

```

The compiler complains: `error CS0165: Use of unassigned local variable 'ws'`

The compiler is Microsoft (R) Visual C# Compiler version 3.10.0-4.21329.37 and the version of .NET Framework is 4.7.2

But, the same program compiles successfully when the `MetaWorkspace` struct definition is moved into a different assembly (DLL) and is included as a reference. How is this so? If this is not a compiler error, am I still dealing with a fully defined `MetaWorkspace` object or have I tread into the undefined zone?

**Comment by spydacarnage** [comment]
The reason why you're getting that error is because you're declaring the local `ws` variable, but you're not explicitly assigning anything to it.  The same applies in the following code, where you are using `x` before explicitly assigning anything to it, even though you would think it has a default value of `0`:
```cs
int x;
string s = x.ToString();
```

On the other hand, if you'd declared your `MetaWorkspace ws;` at the class-level, it does auto-initialise, which is probably what you are seeing when you use it in your second case.

You can see more information on this here: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0165

**Comment by ContentsMayVary** [comment]
@spydacarnage  The question isn't why error CS0165 is generated; the question is why CS0165 _isn't_ generated when `MetaWorkspace` is defined in a different assembly.

**Comment by spydacarnage** [comment]
It doesn't make a difference for me - I get the same error.

Hence why I wondered if, in using it from the separate assembly, the code also changed slightly...

**Reply by sgurukrupa** [reply]
The following program compiles successfully when `MetaWorkspace` is defined in a different assembly.
```
class Program
{
    static void Main(string[] args)
    {
        MetaWorkspace ws;
        ws.Id = "abc";
    }
}
```

**Reply by ufcpp** [reply]
https://github.com/dotnet/roslyn/issues/29319 ?

**Reply by ContentsMayVary** [reply]
Looks like dotnet/roslyn#29319 is indeed the same issue.

**Comment by ContentsMayVary** [comment]
I can reproduce this using .net 5.0 with Visual Studio 16.11.2. When the struct is defined in another assembly I get a warning `Warning CS8887 Use of unassigned local variable 'ws'` and if it's defined in the same assembly I get an error `Error CS0165 Use of unassigned local variable 'ws'`. (Note: I disabled Resharper for this test because it also gives an error/warning.) This question was first asked on StackOverflow: https://stackoverflow.com/questions/69220541/struct-in-a-different-assembly-used-without-initialization#comment122586849_69220541
